docs/tutorials/onnx/fine_tuning_gluon.md

```diff
@@ -0,0 +1,420 @@
+```
+train_root
+__label1
+____image1
+____image2
+__label2
+____image3
+____image4
+```
+
+
+```python
+dataset_train = ImageFolderDataset(root=training_path, transform=transform)
+dataset_test = ImageFolderDataset(root=testing_path, transform=transform)
+```
+
+We use num_workers=Number of CPU cores, which means the dataloading and pre-processing is going to be distributed across multiple processes. This will help preventing our GPU from starving and waiting for the data to be copied across
+
+
+```python
+dataloader_train = DataLoader(dataset_train, batch_size=BATCH_SIZE, last_batch='discard',
+                              shuffle=True, num_workers=NUM_WORKERS)
+dataloader_test = DataLoader(dataset_test, batch_size=BATCH_SIZE, last_batch='discard', 
+                             shuffle=True, num_workers=NUM_WORKERS)
+print("Train dataset: {} images, Test dataset: {} images".format(len(dataset_train), len(dataset_test)))
+```
+
+
+`Train dataset: 6996 images, Test dataset: 1681 images`<!--notebook-skip-line-->
+
+
+
+```python
+categories = dataset_train.synsets
+NUM_CLASSES = len(categories)
+BATCH_SIZE = 32
+```
+
+Let's plot the 1000th image to test the dataset
+
+
+```python
+N = 1000
+plt.imshow(np.transpose(dataset_train[N][0].asnumpy(),(1,2,0)))
+plt.axis('off')
+print(categories[dataset_train[N][1]])
+```
+
+
+`Motorbikes`<!--notebook-skip-line-->
+
+
+
+![png](https://github.com/dmlc/web-data/blob/master/mxnet/doc/tutorials/onnx/motorbike.png?raw=true)<!--notebook-skip-line-->
+
+
+## Fine-Tuning the ONNX model
+
+### Getting the last layer
+
+Load the ONNX model
+
+
+```python
+sym, params = onnx_mxnet.import_model(onnx_path)
+```
+
+This function get the output of a given layer
+
+
+```python
+def get_layer_output(symbol, arg_params, layer_name):
+    all_layers = symbol.get_internals()
+    net = all_layers[layer_name+'_output']
+    net = mx.symbol.Flatten(data=net)
+    new_args = dict({k:arg_params[k] for k in arg_params if k in net.list_arguments()})
+    return (net, new_args)
+```
+
+Here we print the different layers of the network to make it easier to pick the right one
+
+
+```python
+sym.get_internals()
+```
+
+
+
+
+```<Symbol group [input_0, param_0, param_1, convolution0, relu0, lrn0, pad0, pooling0, param_2, param_3, convolution1, relu1, lrn1, pad1, pooling1, param_4, param_5, convolution2, relu2, param_6, param_7, convolution3, relu3, param_8, param_9, convolution4, relu4, pad2, pooling2, _mulscalar0, param_10, param_11, _mulscalar1, fullyconnected0, relu5, _mulscalar2, param_12, param_13, _mulscalar3, fullyconnected1, relu6, _mulscalar4, param_14, param_15, _mulscalar5, fullyconnected2, softmax0]>```<!--notebook-skip-line-->
+
+
+
+We get the network until the output of the `relu6` layer
+
+
+```python
+new_sym, new_params = get_layer_output(sym, params, 'relu6')
+```
+
+### Fine-tuning in gluon
+
+We pick a context
+
+
+```python
+ctx = mx.gpu()
```

Comment:
It will fail if a user doesn't have GPU (like running on mac). For safety reason I recommend to replace it with:
ctx = mx.gpu() if mx.test_utils.list_gpus() else mx.cpu()

Reply:
...And warn a user that using CPU it would take forever to fine tune the model

Reply:
Thanks sounds good, nice little one-liner 👍